src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -73,14 +73,88 @@ public function matchType(Type $patternType, ?Type $flagsType, TrinaryLogic $was
 	 */
 	private function matchRegex(string $regex, ?int $flags, TrinaryLogic $wasMatched): ?Type
 	{
-		$captureGroups = $this->parseGroups($regex);
-		if ($captureGroups === null) {
+		$groupList = $this->parseGroups($regex);
+		if ($groupList === null) {
 			// regex could not be parsed by Hoa/Regex
 			return null;
 		}
 
-		$builder = ConstantArrayTypeBuilder::createEmpty();
+		$trailingOptionals = 0;
+		foreach (array_reverse($groupList) as $captureGroup) {
+			if (!$captureGroup->isOptional()) {
+				break;
+			}
+			$trailingOptionals++;
+		}
+
 		$valueType = $this->getValueType($flags ?? 0);
+		$onlyOptionalTopLevelGroup = $this->getOnlyOptionalTopLevelGroup($groupList);
+		if (
+			$wasMatched->yes()
+			&& $onlyOptionalTopLevelGroup !== null
+		) {
+			// if only one top level capturing optional group exists
+			// we build a more precise constant union of a empty-match and a match with the group
```

Comment:
I am limiting to this specific case for now, because whether $matches contains a offset or not depends on many factors. I did this already for more stuff, but it was easily breaking stuff.

will bring back more sophiscated stuff in future PRs